### ISSUE TITLE
Git for Windows crash when I try to use git difftool -d on a file added with -N option

### ISSUE THREAD

**Issue by sachinrk** [git-for-windows/git#2677]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?
git version 2.27.0.windows.1 
64 bit. 

I recently upgraded the git version, but the issue was reproducible on older versions as well. 

```
$ git --version --build-options

git version 2.27.0.windows.1
cpu: x86_64
built from commit: 907ab1011dce9112700498e034b974ba60f8b407
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.18363.836
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Nothing interesting.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
In a new branch, simply 
notepad++ a.txt
git add -N a.txt
git difftool -d

```
 - What did you expect to occur after running these commands?

No crash, and that git would open my difftool with the diff of the file

 - What actually happened instead?
```
d:\RoutingPlane\sources\dev\CAFE\NanoProxy>git difftool -d
error: unable to read sha1 file of C:\Users\sakul\AppData\Local\Temp\git-difftool.a32148/right/build/configuration/public/xml/a.txt (4b825dc642cb6eb9a060e54bf8d69288fbee4904)
error: could not write 'build/configuration/public/xml/a.txt'
```
Crash dialog saying Git for Windows has stopped working

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Not specific to the repository

**Comment by dscho** [comment]
I can reproduce, even on Linux (read: this problem is _not_ Windows-specific):

```diff
diff --git a/t/t7800-difftool.sh b/t/t7800-difftool.sh
index e66af3682f8..731ed73f421 100755
--- a/t/t7800-difftool.sh
+++ b/t/t7800-difftool.sh
@@ -720,6 +720,14 @@ test_expect_success SYMLINKS 'difftool --dir-diff handles modified symlinks' '
 	test_cmp expect actual
 '
 
+test_expect_success 'add -N and difftool -d' '
+	test_when_finished git reset --hard &&
+
+	test_write_lines A B C >intent-to-add &&
+	git add -N intent-to-add &&
+	git difftool --dir-diff --extcmd ls
+'
+
 test_expect_success 'outside worktree' '
 	echo 1 >1 &&
 	echo 2 >2 &&
```

And this seems to fix it:

```diff
diff --git a/diff.c b/diff.c
index d1ad6a3c4ad..3503dcdf0db 100644
--- a/diff.c
+++ b/diff.c
@@ -5684,6 +5684,9 @@ static void diff_flush_raw(struct diff_filepair *p, struct diff_options *opt)
 	int line_termination = opt->line_termination;
 	int inter_name_termination = line_termination ? '\t' : '\0';
 
+	diff_fill_oid_info(p->one, opt->repo->index);
+	diff_fill_oid_info(p->two, opt->repo->index);
+
 	fprintf(opt->file, "%s", diff_line_prefix(opt));
 	if (!(opt->output_format & DIFF_FORMAT_NAME_STATUS)) {
 		fprintf(opt->file, ":%06o %06o %s ", p->one->mode, p->two->mode,
```

**Comment by sachinrk** [comment]
> I can reproduce, even on Linux (read: this problem is _not_ Windows-specific):
> 
> ```diff
> diff --git a/t/t7800-difftool.sh b/t/t7800-difftool.sh
> index e66af3682f8..731ed73f421 100755
> --- a/t/t7800-difftool.sh
> +++ b/t/t7800-difftool.sh
> @@ -720,6 +720,14 @@ test_expect_success SYMLINKS 'difftool --dir-diff handles modified symlinks' '
>  	test_cmp expect actual
>  '
>  
> +test_expect_success 'add -N and difftool -d' '
> +	test_when_finished git reset --hard &&
> +
> +	test_write_lines A B C >intent-to-add &&
> +	git add -N intent-to-add &&
> +	git difftool --dir-diff --extcmd ls
> +'
> +
>  test_expect_success 'outside worktree' '
>  	echo 1 >1 &&
>  	echo 2 >2 &&
> ```
> 
> And this seems to fix it:
> 
> ```diff
> diff --git a/diff.c b/diff.c
> index d1ad6a3c4ad..3503dcdf0db 100644
> --- a/diff.c
> +++ b/diff.c
> @@ -5684,6 +5684,9 @@ static void diff_flush_raw(struct diff_filepair *p, struct diff_options *opt)
>  	int line_termination = opt->line_termination;
>  	int inter_name_termination = line_termination ? '\t' : '\0';
>  
> +	diff_fill_oid_info(p->one, opt->repo->index);
> +	diff_fill_oid_info(p->two, opt->repo->index);
> +
>  	fprintf(opt->file, "%s", diff_line_prefix(opt));
>  	if (!(opt->output_format & DIFF_FORMAT_NAME_STATUS)) {
>  		fprintf(opt->file, ":%06o %06o %s ", p->one->mode, p->two->mode,
> ```

Thanks for the quick fix! :)

**Comment by dscho** [comment]
It's not actually a fix because it fails the test suite, pointing out a rather big conceptual problem (I did not figure out how to distinguish between a `cache_entry` that is `--intent-to-add` vs one that is simply in need of refreshing).

**Comment by dscho** [comment]
Just a quick update that https://github.com/gitgitgadget/git/pull/654 made it into the `next` branch and will hopefully be included in Git v2.28.0. If not, my plan is to include it in Git _for Windows_ v2.28.0, anyway.

**Comment by sachinrk** [comment]
Sounds great. Thank you for the update.

-Sachin.

**Comment by dscho** [comment]
@sachinrk would you mind testing the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html)?

**Comment by sachinrk** [comment]
Won't mind at all.

 What do I have to do?

**Comment by dscho** [comment]
> Won't mind at all. What do I have to do?

Just install the latest [snapshot](https://wingit.blob.core.windows.net/files/index.html) and re-test whether the bug you reported is fixed or not ;-)

**Comment by sachinrk** [comment]
Can confirm the issue is resolved with the latest git for windows snapshot

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git --version
git version 2.27.0.windows.1.18.g51f9af323b.20200707200701

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>notepad++ a.txt

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git st
On branch user/sakul/entity_body_support
Your branch is ahead of 'origin/user/sakul/entity_body_support' by 1 commit.
  (use "git push" to publish your local commits)

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        a.txt

nothing added to commit but untracked files present (use "git add" to track)

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git add -N a.txt

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git st
On branch user/sakul/entity_body_support
Your branch is ahead of 'origin/user/sakul/entity_body_support' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        new file:   a.txt

no changes added to commit (use "git add" and/or "git commit -a")

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git difftool -d

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>git difftool -d

d:\RoutingPlane\sources\test\CAFE\NanoProxyTests>

**Comment by sachinrk** [comment]
Is it possible to port this fix to git used with vfs? specifically - 2.23.0.vfs.1.1.

**Comment by dscho** [comment]
@sachinrk I believe that [the latest VFS release](https://github.com/microsoft/VFSForGit/releases/tag/v1.0.20210.1) comes with this fix.
